File: android/app/src/main/java/com/xirigo/ecommerce/core/designsystem/component/XGPriceText.kt
package com.xirigo.ecommerce.core.designsystem.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.sp
import com.xirigo.ecommerce.R
import com.xirigo.ecommerce.core.designsystem.theme.PoppinsFontFamily
import com.xirigo.ecommerce.core.designsystem.theme.SourceSans3FontFamily
import com.xirigo.ecommerce.core.designsystem.theme.XGColors
import com.xirigo.ecommerce.core.designsystem.theme.XGSpacing
import com.xirigo.ecommerce.core.designsystem.theme.XGTheme

// Token source: components/atoms/xg-price-text.json — variant font sizes
private val DefaultCurrencyFontSize = 22.78.sp
private val DefaultIntegerFontSize = 27.33.sp
private val DefaultDecimalFontSize = 18.98.sp
private val StandardCurrencyFontSize = 20.sp
private val StandardIntegerFontSize = 20.sp
private val StandardDecimalFontSize = 14.sp
private val SmallCurrencyFontSize = 14.sp
private val SmallIntegerFontSize = 18.sp
private val SmallDecimalFontSize = 14.sp

// Token source: components/atoms/xg-price-text.json — strikethrough font sizes
private val DefaultStrikethroughFontSize = 15.18.sp
private val StandardStrikethroughFontSize = 14.sp

/** Visual style variants for price display. Maps to XGPriceStyle in the token spec. */
enum class XGPriceStyle {
    /** Default discount price (Featured/Popular cards). */
    Default,

    /** Standard variant for New Arrivals cards (20/20/14). */
    Standard,

    /** Small variant for compact layouts. */
    Small,

    /** Deal variant — brand secondary color for daily deal contexts. */
    Deal,
}

/** Formatted price display with currency symbol, integer, and decimal parts.
 *  When [price] is null the component renders nothing (hides entirely). */
@Composable
fun XGPriceText(
    price: String?,
    modifier: Modifier = Modifier,
    originalPrice: String? = null,
    currencySymbol: String = "\u20AC",
    style: XGPriceStyle = XGPriceStyle.Default,
    strikethroughFontSize: Float = DefaultStrikethroughFontSize.value,
    layout: XGPriceLayout = XGPriceLayout.Inline,
) {
    // Null price fallback: hide component entirely (DQ-10)
    if (price == null) return

    val priceColor: Color
    val currencyFontSize: Float
    val integerFontSize: Float
    val decimalFontSize: Float

    when (style) {
        XGPriceStyle.Default -> {
            priceColor = XGColors.PriceSale
            currencyFontSize = DefaultCurrencyFontSize.value
            integerFontSize = DefaultIntegerFontSize.value
            decimalFontSize = DefaultDecimalFontSize.value
        }
        XGPriceStyle.Standard -> {
            priceColor = XGColors.PriceSale
            currencyFontSize = StandardCurrencyFontSize.value
            integerFontSize = StandardIntegerFontSize.value
            decimalFontSize = StandardDecimalFontSize.value
        }
        XGPriceStyle.Small -> {
            priceColor = XGColors.PriceSale
            currencyFontSize = SmallCurrencyFontSize.value
            integerFontSize = SmallIntegerFontSize.value
            decimalFontSize = SmallDecimalFontSize.value
        }
        XGPriceStyle.Deal -> {
            priceColor = XGColors.BrandSecondary
            currencyFontSize = DefaultCurrencyFontSize.value
            integerFontSize = DefaultIntegerFontSize.value
            decimalFontSize = DefaultDecimalFontSize.value
        }
    }

    val parts = splitPrice(price)

    val formattedPrice = "$currencySymbol${parts.first}.${parts.second}"
    val formattedOriginalPrice = originalPrice?.let {
        val origParts = splitPrice(it)
        "$currencySymbol${origParts.first}.${origParts.second}"
    }

    val accessibilityDescription = if (formattedOriginalPrice != null) {
        stringResource(R.string.common_sale_price_label, formattedPrice, formattedOriginalPrice)
    } else {
        stringResource(R.string.common_price_label, formattedPrice)
    }

    val compositePriceText = @Composable {
        Text(
            text = buildAnnotatedString {
                withStyle(
                    SpanStyle(
                        fontFamily = SourceSans3FontFamily,
                        fontWeight = FontWeight.Black,
                        fontSize = currencyFontSize.sp,
                        color = priceColor,
                    ),
                ) {
                    append(currencySymbol)
                }
                withStyle(
                    SpanStyle(
                        fontFamily = SourceSans3FontFamily,
                        fontWeight = FontWeight.Black,
                        fontSize = integerFontSize.sp,
                        color = priceColor,
                    ),
                ) {
                    append(parts.first)
                }
                withStyle(
                    SpanStyle(
                        fontFamily = SourceSans3FontFamily,
                        fontWeight = FontWeight.Black,
                        fontSize = decimalFontSize.sp,
                        color = priceColor,
                    ),
                ) {
                    append(",${parts.second}")
                }
            },
        )
    }

    val strikethroughText = @Composable {
        if (originalPrice != null) {
            val origParts = splitPrice(originalPrice)
            Text(
                text = "$currencySymbol${origParts.first},${origParts.second}",
                style = TextStyle(
                    fontFamily = PoppinsFontFamily,
                    fontWeight = FontWeight.Medium,
                    fontSize = strikethroughFontSize.sp,
                    textDecoration = TextDecoration.LineThrough,
                    color = XGColors.PriceStrikethrough,
                ),
            )
        }
    }

    when (layout) {
        XGPriceLayout.Inline -> {
            Row(
                modifier = modifier.semantics { contentDescription = accessibilityDescription },
                horizontalArrangement = Arrangement.spacedBy(XGSpacing.SM),
                verticalAlignment = Alignment.Bottom,
            ) {
                compositePriceText()
                strikethroughText()
            }
        }
        XGPriceLayout.Stacked -> {
            Column(
                modifier = modifier.semantics { contentDescription = accessibilityDescription },
            ) {
                strikethroughText()
                compositePriceText()
            }
        }
    }
}

private fun splitPrice(price: String): Pair<String, String> {
    val cleaned = price.replace("[^0-9.,]".toRegex(), "")
    val separator = when {
        cleaned.contains(",") -> ","
        cleaned.contains(".") -> "."
        else -> return Pair(cleaned, "00")
    }
    val parts = cleaned.split(separator)
    val integer = parts.getOrElse(0) { "0" }
    val decimal = parts.getOrElse(1) { "00" }.take(2).padEnd(2, '0')
    return Pair(integer, decimal)
}

private object PreviewPrices {
    const val REGULAR = "29.99"
    const val ORIGINAL = "39.99"
    const val SMALL = "9.99"
    const val DEAL = "89.99"
    const val DEAL_ORIGINAL = "149.99"
    const val STANDARD_STRIKETHROUGH = 14f
}

@Preview(showBackground = true)
@Composable
private fun XGPriceTextDefaultPreview() {
    XGTheme {
        XGPriceText(price = PreviewPrices.REGULAR)
    }
}

@Preview(showBackground = true)
@Composable
private fun XGPriceTextSalePreview() {
    XGTheme {
        XGPriceText(price = PreviewPrices.REGULAR, originalPrice = PreviewPrices.ORIGINAL)
    }
}

@Preview(showBackground = true)
@Composable
private fun XGPriceTextStandardPreview() {
    XGTheme {
        XGPriceText(
            price = PreviewPrices.REGULAR,
            originalPrice = PreviewPrices.ORIGINAL,
            style = XGPriceStyle.Standard,
            strikethroughFontSize = PreviewPrices.STANDARD_STRIKETHROUGH,
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun XGPriceTextSmallPreview() {
    XGTheme {
        XGPriceText(price = PreviewPrices.SMALL, style = XGPriceStyle.Small)
    }
}

@Preview(showBackground = true)
@Composable
private fun XGPriceTextDealPreview() {
    XGTheme {
        XGPriceText(
            price = PreviewPrices.DEAL,
            originalPrice = PreviewPrices.DEAL_ORIGINAL,
            style = XGPriceStyle.Deal,
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun XGPriceTextStackedPreview() {
    XGTheme {
        XGPriceText(
            price = PreviewPrices.DEAL,
            originalPrice = PreviewPrices.DEAL_ORIGINAL,
            layout = XGPriceLayout.Stacked,
        )
    }
}

@Preview(showBackground = true, name = "XGPriceText Null Price")
@Composable
private fun XGPriceTextNullPricePreview() {
    XGTheme {
        Column {
            Text("Before price:")
            XGPriceText(price = null)
            Text("After price (nothing between):")
        }
    }
}

File: ios/XiriGoEcommerce/Core/DesignSystem/Component/XGPriceText.swift
import SwiftUI

// MARK: - XGPriceStyle

/// Visual style variants for price display.
/// Token source: `components/atoms/xg-price-text.json`.
enum XGPriceStyle {
    /// Default discount price — brand primary color (Featured/Popular cards).
    case `default`
    /// Standard variant for New Arrivals / Standard cards (20/20/14).
    case standard
    /// Small variant for compact layouts (product cards).
    case small
    /// Deal variant — brand secondary color for daily deal contexts.
    case deal

    // MARK: - Internal

    var currencyFontSize: CGFloat {
        switch self {
            case .deal,
                 .default:
                Constants.defaultCurrencyFontSize
            case .standard:
                Constants.standardCurrencyFontSize
            case .small:
                Constants.smallCurrencyFontSize
        }
    }

    var integerFontSize: CGFloat {
        switch self {
            case .deal,
                 .default:
                Constants.defaultIntegerFontSize
            case .standard:
                Constants.standardIntegerFontSize
            case .small:
                Constants.smallIntegerFontSize
        }
    }

    var decimalFontSize: CGFloat {
        switch self {
            case .deal,
                 .default:
                Constants.defaultDecimalFontSize
            case .standard:
                Constants.standardDecimalFontSize
            case .small:
                Constants.smallDecimalFontSize
        }
    }

    var color: Color {
        switch self {
            case .default,
                 .small,
                 .standard:
                XGColors.priceSale
            case .deal:
                XGColors.brandSecondary
        }
    }

    // MARK: - Private

    private enum Constants {
        static let defaultCurrencyFontSize: CGFloat = 22.78
        static let defaultIntegerFontSize: CGFloat = 27.33
        static let defaultDecimalFontSize: CGFloat = 18.98

        static let standardCurrencyFontSize: CGFloat = 20
        static let standardIntegerFontSize: CGFloat = 20
        static let standardDecimalFontSize: CGFloat = 14

        static let smallCurrencyFontSize: CGFloat = 14
        static let smallIntegerFontSize: CGFloat = 18
        static let smallDecimalFontSize: CGFloat = 14
    }
}

// MARK: - XGPriceLayout

/// Controls the arrangement of sale price and strikethrough in ``XGPriceText``.
enum XGPriceLayout {
    /// Sale price + strikethrough side-by-side (HStack). Default for standard/grid cards.
    case inline
    /// Strikethrough above, sale price below (VStack). Used for featured/horizontal-scroll cards.
    case stacked
}

// MARK: - XGPriceText

/// Three-part composite price display component.
/// Renders currency symbol, integer part, and decimal part with separate font sizes.
/// Uses Source Sans 3 Black font per design tokens.
///
/// Token source: `components/atoms/xg-price-text.json`.
struct XGPriceText: View {
    // MARK: - Lifecycle

    init(
        price: String?,
        originalPrice: String? = nil,
        currencySymbol: String = "\u{20AC}",
        style: XGPriceStyle = .default,
        strikethroughFontSize: CGFloat = Constants.defaultStrikethroughFontSize,
        layout: XGPriceLayout = .inline,
    ) {
        self.price = price
        self.originalPrice = originalPrice
        self.currencySymbol = currencySymbol
        self.style = style
        self.strikethroughFontSize = strikethroughFontSize
        self.layout = layout
    }

    // MARK: - Internal

    var body: some View {
        // Null price fallback: hide component entirely (DQ-10)
        if price != nil {
            Group {
                switch layout {
                    case .inline:
                        HStack(alignment: .firstTextBaseline, spacing: XGSpacing.sm) {
                            compositePrice

                            if let originalPrice {
                                strikethroughPrice(originalPrice)
                            }
                        }

                    case .stacked:
                        VStack(alignment: .leading, spacing: XGSpacing.xxs) {
                            if let originalPrice {
                                strikethroughPrice(originalPrice)
                            }

                            compositePrice
                        }
                }
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(accessibilityDescription)
        }
    }

    // MARK: - Private

    private enum Constants {
        static let defaultStrikethroughFontSize: CGFloat = 15.18
        static let standardStrikethroughFontSize: CGFloat = 14
    }

    private let price: String?
    private let originalPrice: String?
    private let currencySymbol: String
    private let style: XGPriceStyle
    private let strikethroughFontSize: CGFloat
    private let layout: XGPriceLayout

    private var hasSale: Bool {
        originalPrice != nil
    }

    private var priceColor: Color {
        style.color
    }

    private var priceParts: (integer: String, decimal: String) {
        let priceValue = price ?? "0"
        let components = priceValue.split(separator: ".", maxSplits: 1)
        let integer = String(components.first ?? "0")
        let decimal = components.count > 1 ? "," + String(components.last ?? "00") : ""
        return (integer, decimal)
    }

    private var accessibilityDescription: String {
        let priceValue = price ?? "0"
        if let originalPrice {
            return String(
                localized: "common_sale_price_label \(currencySymbol) \(priceValue) \(currencySymbol) \(originalPrice)",
            )
        }
        return String(
            localized: "common_price_label \(currencySymbol) \(priceValue)",
        )
    }

    private var compositePrice: some View {
        HStack(alignment: .firstTextBaseline, spacing: 0) {
            Text(currencySymbol)
                .font(priceFont(size: style.currencyFontSize))
                .foregroundStyle(priceColor)

            Text(priceParts.integer)
                .font(priceFont(size: style.integerFontSize))
                .foregroundStyle(priceColor)

            Text(priceParts.decimal)
                .font(priceFont(size: style.decimalFontSize))
                .foregroundStyle(priceColor)
        }
    }

    private func strikethroughPrice(_ originalPrice: String) -> some View {
        Text(currencySymbol + originalPrice)
            .font(XGTypography.strikethroughFont(size: strikethroughFontSize))
            .foregroundStyle(XGColors.priceStrikethrough)
            .strikethrough()
    }

    private func priceFont(size: CGFloat) -> Font {
        XGTypography.priceFont(size: size)
    }
}

// MARK: - Previews

#Preview("XGPriceText Default") {
    VStack(alignment: .leading, spacing: XGSpacing.sm) {
        XGPriceText(price: "29.99", style: .small)
        XGPriceText(price: "149.99")
        XGPriceText(price: "99.99", style: .standard)
        XGPriceText(price: "89.99", style: .deal)
    }
    .padding()
    .xgTheme()
}

#Preview("XGPriceText Sale") {
    VStack(alignment: .leading, spacing: XGSpacing.sm) {
        XGPriceText(price: "19.99", originalPrice: "29.99", style: .small)
        XGPriceText(price: "89.99", originalPrice: "149.99")
        XGPriceText(price: "49.99", originalPrice: "69.99", style: .standard, strikethroughFontSize: 14)
        XGPriceText(price: "49.99", originalPrice: "99.99", style: .deal)
    }
    .padding()
    .xgTheme()
}

#Preview("XGPriceText Stacked") {
    VStack(alignment: .leading, spacing: XGSpacing.sm) {
        XGPriceText(price: "89.99", originalPrice: "149.99", layout: .stacked)
        XGPriceText(price: "49.99", originalPrice: "69.99", style: .standard, layout: .stacked)
    }
    .padding()
    .xgTheme()
}

#Preview("XGPriceText Nil Price") {
    VStack(alignment: .leading, spacing: XGSpacing.sm) {
        Text("Before price:")
        XGPriceText(price: nil)
        Text("After price (nothing between):")
    }
    .padding()
    .xgTheme()
}
